mqtt/mqtt-bridge/src/persist/mod.rs

```diff
@@ -0,0 +1,49 @@
+use async_trait::async_trait;
+use futures_util::stream::Stream;
+use mqtt3::proto::Publication;
+use thiserror::Error;
+
+mod memory;
+
+// Persistence used in bridge.
+#[async_trait]
+trait Persist<'a> {
+    type Loader: Stream;
+
+    fn new() -> Self;
+
+    async fn push(&mut self, message: Publication) -> Result<Key, PersistError>;
+
+    async fn remove(&mut self, key: Key) -> Result<bool, PersistError>;
+
+    async fn loader(&'a mut self, batch_size: usize) -> Self::Loader;
+}
+
+// Keys used in persistence.
+// Ordered by offset
+#[derive(Hash, Eq, Ord, PartialOrd, PartialEq, Clone, Debug)]
+pub struct Key {
+    offset: u32,
+}
+
+// TODO REVIEW: these generic error types don't wrap impl specific types
+#[derive(Debug, Error)]
+pub enum PersistError {
+    #[error("Failed to remove messages from persistence")]
+    Removal(),
+}
```

Comment:
What is the best practice here for handling impl specific errors?

Reply:
I still have this question, but it isn't relevant anymore. In handling another comment, I converted the remove return type to option. Now we don't need this.